src/rules.jl

```diff
@@ -131,6 +131,7 @@ init(o::RMSProp, x::AbstractArray) = (zero(x), o.centred ? zero(x) : false)
 
 function apply!(o::RMSProp, state, x::AbstractArray{T}, dx) where T
   η, ρ, ϵ = T(o.eta), T(o.rho), T(o.epsilon)
+  ϵ = max(ϵ, eps(T(0)))
```

Comment:
I don't think this does the right thing:
```
julia> eps(Float32)  # roughly what we hard-code now
1.1920929f-7

julia> eps(0f0)  # what this PR uses
1.0f-45
```
I think a better pattern is this:
```suggestion
  η, ρ, ϵ = T(o.eta), T(o.rho), max(T(o.epsilon), eps(T))
```

Reply:
I chose `eps(T(0))` because i didn't want to change the current behavior `eps=1f-8` for `Float32` but do the minimal change needed for Float16
```julia
julia> eps(Float16)
Float16(0.000977)

julia> eps(Float16(0))
Float16(6.0e-8)
```

Reply:
This doesn't seem big enough to avoid the issue in https://github.com/FluxML/Optimisers.jl/issues/167 : 
```julia
julia> using Optimisers

julia> x = Float16[0.579, -0.729, 0.5493]
3-element Vector{Float16}:
  0.579
 -0.729
  0.5493

julia> δx = Float16[-0.001497, 0.0001875, -0.013176]
3-element Vector{Float16}:
 -0.001497
  0.0001875
 -0.013176

julia> os = Optimisers.setup(Optimisers.Adam(Float16(1e-4)), x);

julia> os, x = Optimisers.update(os, x, δx)
(Leaf(Adam(0.000100017, (0.9, 0.999), 1.0e-8), (Float16[-0.0001498, 1.88e-5, -0.001319], Float16[0.0, 0.0, 2.0e-7], (0.81, 0.998))), Float16[Inf, -Inf, 0.5493])

julia> os = Optimisers.setup(Optimisers.Adam(eta=1e-4, epsilon=eps(Float16(0))), x);  # this PR

julia> os, x = Optimisers.update(os, x, δx)
(Leaf(Adam(0.0001, (0.9, 0.999), 5.96046e-8), (Float16[-0.0001498, 1.88e-5, -0.001319], Float16[0.0, 0.0, 2.0e-7], (0.81, 0.998))), Float16[Inf, -Inf, 0.5522])

julia> os = Optimisers.setup(Optimisers.Adam(eta=1e-4, epsilon=eps(Float16)), x);  # as suggested in #167

julia> os, x = Optimisers.update(os, x, δx)
(Leaf(Adam(0.0001, (0.9, 0.999), 0.000976562), (Float16[-0.0001498, 1.88e-5, -0.001319], Float16[0.0, 0.0, 2.0e-7], (0.81, 0.998))), Float16[Inf, -Inf, 0.5522])
```
... in fact neither are big enough. 

So, what problem do we want this to solve?

Reply:
Ok. How broadly is 6e-8 big enough for this purpose? It seems a little random... maybe I'd be happier to hard-code an obviously chosen number, which is what `1e-8` is... like `_eps(::Type{Float16}, e) = max(1e-7, e)`?

Reply:
done